### ISSUE TITLE
Union of tuples not recognised properly

### ISSUE THREAD

**Issue by Rizhiy** [microsoft/pyright#3919]
**Describe the bug**
When union of tuples is specified as return type, connection between tuple member is not detected.

**To Reproduce**
```python
from typing import Literal, Union


def foo(choice: bool) -> Union[tuple[float, float], tuple[None, None]]:
    if choice:
        return 1.0, 2.0
    return None, None


a, b = foo(True)
if a is not None:
    print(b + 1.0)

if isinstance(a, float):
    print(b + 1.0)


def bar(choice: bool) -> Union[tuple[Literal[1], float], tuple[Literal[2], str]]:
    if choice:
        return 1, 1.0
    return 2, "bar"


a, b = bar(True)
if a == 1:
    print(b + 1.0)
if a == 2:
    print(b + "1.0")
```
Produces errors:
```
  /home/rizhiy/projects/binance-test/tmp.py:12:11 - error: Operator "+" not supported for "None" (reportOptionalOperand)
  /home/rizhiy/projects/binance-test/tmp.py:15:11 - error: Operator "+" not supported for "None" (reportOptionalOperand)
  /home/rizhiy/projects/binance-test/tmp.py:26:11 - error: Operator "+" not supported for types "float | str" and "float"
    Operator "+" not supported for types "str" and "float" when expected type is "object" (reportGeneralTypeIssues)
  /home/rizhiy/projects/binance-test/tmp.py:28:11 - error: Operator "+" not supported for types "float | str" and "Literal['1.0']"
    Operator "+" not supported for types "float" and "Literal['1.0']" when expected type is "object" (reportGeneralTypeIssues)
```
Last error makes no sense in particular.

**Expected behavior**
No errors

**Comment by erictraut** [comment]
Yes, this is by design. Type narrowing operates on individual expressions. It it's able to conditionally narrow based on "entanglement" between different expressions, like `a` and `b` in your example above. Tracking conditional narrowing across different expressions would be computationally infeasible, and I don't know of any type checker or any compiler in any language that does this.

If you want to narrow the type of `b` in an expression, you will need to use a type guard expression directly on `b`. You can't rely on a type guard that operates on `a` to narrow the type of `b`.
